### DISCUSSION TITLE
Exception when signing a PDF with filled forms

### DISCUSSION THREAD

**Discussion by fredericoschardong** [MatthiasValvekens/pyHanko#49] [General]
Hey Matthias,

I have been playing arround the subject of our last discussion: signing filled forms, but hit a roadblock. Here it is:

**Describe the bug**
I get `pyhanko.pdf_utils.misc.PdfReadError` when signing a pdf with any filled form I could find. Perhaps the forms are not A-grade and are the root cause of the issue, or the programs I am using to fill them (okular and evince). I can sign any pdf with forms just fine if and only if they are empty.

**To Reproduce**
1. Download [example.zip](https://github.com/MatthiasValvekens/pyHanko/files/7297587/example.zip) 
2. Fill the form.pdf and save the resulting pdf under the same name
3. Run `python3 sign.py` to sign the filled form and get an error like `pyhanko.pdf_utils.misc.PdfReadError: Generation 0 of object 3 was never freed, but reused later.`

**Expected behavior**
No errors and a signed pdf.

**Environment (please complete the following information):**
 - OS: ubuntu
 - Version: 21.04

**Comment by MatthiasValvekens** [comment]
Hmmm, this is a though one.

First, some theoretical analysis/rambling: I guess Okular and evince use the same PDF writing backend, then. I tried filling out the form with Okular, and the revision made by Okular overrides the form field's appearance stream while also incrementing the generation number. This is a misinterpretation of what generation numbers are for: when overriding an existing object, you can simply clobber the previous incarnation without bumping the gen number --- in fact Okular does exactly that when overriding the form field itself. You only need to start a new generation if you're reusing the object ID of a _previously freed_ object. However, since object IDs are not in short supply, pretty much no-one does that these days.

PyHanko's cross-reference table parsing logic (which was written with forensic analysis on signed files in mind) treats this as an error because the spec doesn't tell you what to do when there are multiple "currently active"/unfreed generations with the same object ID in a file, and different implementations do different things there. This can be abused to craft a PDF file that renders differently with different viewers, which I'm sure you'll agree is a problem when dealing with signed files ;).

Anyway, the whole object freeing/generation number thing is confusingly specified in general, so I concede that my own interpretation may very well be wrong. The issues with differing interpretations "in the wild" are real, though.

**TL;DR**: The form file looks fine. This is probably a bug in Okular/evince/...'s form filling logic (which is an interesting data point in itself).

----------

So, what can we do about that? Whether anything needs to be done depends on what you want to achieve:

 - PyHanko never writes nonzero generation numbers and never frees objects, so any putative form filling implementation in pyHanko would not be exposed to this class of bugs. If you were intending to use pyHanko if/when it gains generic form filling capabilities, this won't be an issue. :)
 - If you need a way to test signing behaviour with filled forms in the meantime, there are other viewers out there that support form filling: you could use Adobe Reader (or Acrobat, obviously), Foxit's viewer, or even your browser's built-in PDF viewer (Firefox, Chrome and their derivatives should be OK).
 - I could also try to figure out whether it's possible to add a flag to the xref parser to ignore this sort of quirks, but I can't say off-hand whether that will be easy. It would also have to be disabled by default, for the reasons that I mentioned above.

Anyway, that turned out quite a bit longer than I anticipated... Thanks for the test files, and let me know what you think.

**Comment by fredericoschardong** [comment]
Thank you for the explanation.

> could also try to figure out whether it's possible to add a flag to the xref parser to ignore this sort of quirks, but I can't say off-hand whether that will be easy. It would also have to be disabled by default, for the reasons that I mentioned above.

Nah, don't worry about it. Indeed I was able to sign the filled form using adobe reader. Interestingly enough, if I use Chrome, it fails in the same manner as Okular, and it worked as expected using Firefox.

Nonetheless, I couldn't get the expected results when setting the second signature to `docmdp_permissions=MDPPerm.NO_CHANGES`. Adobe reader still allows changes to the form; please see [example2.zip](https://github.com/MatthiasValvekens/pyHanko/files/7307986/example2.zip) for a working example. Here is the output on adobe reader asking to fill the (already filled) form:

![output](https://user-images.githubusercontent.com/3732628/136487436-33886af5-f220-48a0-ad1c-2a41a0bc4973.png)

Am I missing something on PyHanko's side, or is my understanding of expecting the reader not to allow any more changes wrong?

**Comment by MatthiasValvekens** [comment]
That's a valid question. The logic for DocMDP policies on an approval signature (i.e. a signature that's not a certification signature) is a bit wonky, and in case of `NO_CHANGES`, I should probably tighten up this piece of fallback logic that's used when there's no field lock dictionary on the field: https://github.com/MatthiasValvekens/pyHanko/blob/6eef056fb5ac32f87fedc9ae8dc1181c3a45adac/pyhanko/sign/signers/pdf_signer.py#L1288-L1295. I guess it makes sense to call that a bug.

In the meantime, you can try creating a signature field with a field lock dictionary that locks everything, and then sign that. Here's the documentation on how to do field locking in pyHanko: https://pyhanko.readthedocs.io/en/latest/lib-guide/sig-fields.html#document-modification-policy-settings. You can combine that with a DocMDP policy setting, which is very common, but it's a PDF 2.0 thing, and it's pretty poorly specified in general (it's a common source of complaints in the industry).

Also, in case you're not already doing that: in multi-signer workflows, it's good practice to create all visible signature fields before filling them. The rationale behind that is that the initial signer should have the right to know where all the other signatures will go, so they can make sure that the visual appearances of later signatures won't change the meaning of any document content. Not all validators enforce that principle, though. PyHanko's only does it by default if a certification signature is present.

So the ideal workflow is more or less this:

 - Set up all form fields, with field lock dictionaries for the signature fields if necessary. You can put a DocMDP `NO_CHANGES` trigger on the field that will be signed last (subject to the caveats from above)
 - Optionally put in a visible or invisible certification signature with DocMDP set to "form filling" (if that makes sense in your context)
 - Do form filling operations
 - Put in the "regular" signatures

I can't do any real testing right now, but modulo some details I think this should work. :)

**Reply by MatthiasValvekens** [reply]
Upon inspection, this is broader than I anticipated. The DocMDP perms in the `PdfSignatureMetadata` object are only used when `certify=True`, and there's no easy way to turn that into "implicitly apply DocMDP using a field lock" without breaking stuff, since the default is not `None` but `FORM_FILLING`.

Fixing this properly definitely requires an API change, but I'm not sure what the cleanest way would be just yet. I'll create a separate issue for that.

**Reply by MatthiasValvekens** [reply]
See #50.

**Comment by fredericoschardong** [comment]
I guess my new roadblock has to do with what you said about the field lock. [Here](https://github.com/MatthiasValvekens/pyHanko/files/7422062/test.zip) is the code. Starting from a signed empty form, which is then filled with pdfbox, everything works if I sign it with the high-level API (see sign_2.py). However, I can't get a correct result (according to Adobe Reader) using the low-level API (see sign_2_error.py).

The low-level API leaves the field lock untouched, is that breaking things?

**Reply by MatthiasValvekens** [reply]
Wait, I'm a little confused now. You have this at the end of your signing routine:

```
    with open('filled form signed again error.pdf', 'wb') as outf:
        outf.write(output.getbuffer())
        sig_contents = cms_writer.send(cms_bytes)
```

But that won't do what you think it does: it'll dump the contents of the in-memory representation of the PDF file to disk, and then write the signature to the in-memory buffer. In other words, you end up with a placeholder value in the final document instead of a valid signature.

That should be

```
    sig_contents = cms_writer.send(cms_bytes)
    with open('filled form signed again error.pdf', 'wb') as outf:
        outf.write(output.getbuffer())
```

Or am I missing the point?

**Reply by fredericoschardong** [reply]
You are right. I misplaced that line. Nonetheless, after fixing it, I am still facing a strange issue: Adobe Reader inverts the order of the signatures, while Okular does not. That only happens using the lower-level API. 

As Adobe Reader puts the second signature (which locks fields) first, my little scheme collapses. Do you have any idea why it might be doing that?

**Reply by MatthiasValvekens** [reply]
Ooh, now I see. You're passing `certify=True` in the second signature, and you can't do that according to the standard: a certification signature always has to go first (which is probably also why the order is switched up in the Acrobat UI). I should probably put logic into pyHanko that prevents you from making that mistake, as it's a pretty huge foot-gun right now.

The only way to perform any field locking in your situation is to add a field lock dictionary to the field containing the last signature. There's documentation on how to do that [here](https://pyhanko.readthedocs.io/en/latest/lib-guide/sig-fields.html#document-modification-policy-settings). Preferably, you should add the lock dictionary to the last field to be signed _before_ producing the first signature, though.

**Reply by fredericoschardong** [reply]
Thank you, Matthias. I've made the changes you pointed out and everything works now.

**Reply by MatthiasValvekens** [reply]
Great! Yesterday, I put in an extra check that will cause the (high-level) signing API to fail loudly in this situation.
Unfortunately, doing the same thing in the low-level signing logic cleanly is a bit trickier, but that's part of the tradeoff, I guess.

**Reply by mmdanny89** [reply]
hi, but in both examples when run sign validations on PDF file  that have two fields signs, the first sign is invalid:
```
with open('filled form signed again.pdf', 'rb') as doc:
        r = PdfFileReader(
        vc = ValidationContext()
        sig = r.embedded_signatures[0]
        #sig = r.embedded_signatures[1]
        status = validate_pdf_signature(sig, vc)
        print(status.pretty_print_details())
```
**when is this sig = r.embedded_signatures[0],  the result is:** 
```
Modifications
-------------
The signature does not cover the entire file.
Some modifications may be illegitimate, and they appear to be incompatible with the current document modification policy.
```
**when is this sig = r.embedded_signatures[1],  the result is:** 
```
Modifications
-------------
The signature covers the entire file.
```

**Reply by MatthiasValvekens** [reply]
Hi @mmdanny89,

This kind of issue is next to impossible to diagnose effectively without the actual document to look at. :)

Could you share the signed file somewhere?

**Reply by mmdanny89** [reply]
yes, sure:  [this example](https://github.com/MatthiasValvekens/pyHanko/files/7422062/test.zip) pdf file is: _filled form signed again.pdf_.
In library tests i can´t see any test for add more one sign on same pdf,  (open file,  sign, close file; and open again and .. sign again and...), you may help with this?, thanks

**Reply by mmdanny89** [reply]
the certificate chain is valid, CA Root, CA Intermediarie and Certificate .p12

**Reply by MatthiasValvekens** [reply]
Hi @mmdanny89,

Thanks for the test files. Running `pyhanko --verbose sign validate filled\ form\ signed.pdf` reveals that pyHanko actually already considers the edits after the first signature to be problematic, due to the fact that one of the form dictionaries gained an extra key. The key in question is pretty harmless, so that's arguably a bug in the validator. It's not a hard one to fix, so I'll deal with it.

I'm not sure what relation the `filled form signed again error.pdf` file has to the other documents, but in that one the second signature consists entirely of zeroes.

> In library tests i can´t see any test for add more one sign on same pdf, (open file, sign, close file; and open again and .. sign again and...), you may help with this?, thanks

There are many such tests :). Most of them can be found in `test_diff_analysis.py`, but some are also in `test_pades.py` and `test_signing.py`.

**Reply by MatthiasValvekens** [reply]
Oh, by the way, you can work around the bug by adding `/Type` to `VALUE_UPDATE_KEYS` in `pyhanko.sign.diff_analysis.constants` at runtime. If you prefer not messing with constants, it's also possible to instantiate your own diff policy with a custom `value_update_keys` parameter, but that's a bit more complicated.

**Reply by mmdanny89** [reply]
Thanks @MatthiasValvekens

**Reply by MatthiasValvekens** [reply]
There are apparently some other issues with form fields in your file, I'm looking into those now.

**Reply by MatthiasValvekens** [reply]
Seems that the form filling operation applied to the file runs into this limitation of the validator: https://github.com/MatthiasValvekens/pyHanko/blob/baa4e1a4fb2a64281cf6c83c77715b01baf060a5/pyhanko/sign/diff_analysis/rules/form_field_rules.py#L684-L689. The bad news is that this one is more difficult to apply a correct fix for. The good news is that the reference counter mechanism mentioned in the `TODO` comment has actually been added in the meantime.

Another weird thing is that the root of the page tree is overridden in the update, which is usually a bad thing, but in this case the update actually doesn't change anything. Maybe I should also add a "pointless update rule" to permit updates that _literally_ leave the object they replace exactly the same.

-----
 
Anyway, both of these quirks will require some time to fix, but it's been a while since I got complaints about the diff validator, so good find! Do you happen to know which software was used to fill in the form fields?

**Reply by mmdanny89** [reply]
mmm, no.. but, this sound good, i maked new test with libreoffice -> export pdf , with two signs and validator work,

**Reply by mmdanny89** [reply]
both signs are valid, but this issue of modification is live.. OK, Thanks @MatthiasValvekens. Have a good week.
